cm/master.py

```diff
@@ -1463,6 +1386,111 @@ def _add_data_fs(fs_name=None):
+        elif self.app.config.filesystem_templates: # Legacy behavior - process filesystem_templates directly or snaps.yaml
+            self.process_filesystem_templates(cluster_type, pss, storage_type, self.app.config.filesystem_templates)
+        else:
+            if cluster_type != "Test":
+                log.error("No filesystem templates defined for cluster type: {0}".format(cluster_type))
+
+    # TODO: Remove special case handling for Galaxy filesystem - requires change in UI and initialisation logic
+    def process_filesystem_templates(self, cluster_type, pss, storage_type, filesystem_templates):
+        # Turn those data sources into file systems
+        if filesystem_templates:
+            attached_volumes = self.get_attached_volumes()
+            for fs_template in [s for s in filesystem_templates if 'name' in s]:
+                log.debug("Processing filesystem template: {0}".format(fs_template['name']))
+                fs = Filesystem(self.app, fs_template['name'],
+                                svc_roles=ServiceRole.from_string_array(fs_template.get('roles', None)))
+                # Check if an already attached volume maps to the current filesystem
+                att_vol = self.get_vol_if_fs(attached_volumes, fs_template['name'])
+                if att_vol:
+                    log.debug("{0} file system has volume(s) already attached".format(
+                        fs_template['name']))
+                    fs.add_volume(vol_id=att_vol.id,
+                                  size=att_vol.size, from_snapshot_id=att_vol.snapshot_id)
+                elif 'snap_id' in fs_template:
+                    log.debug("There are no volumes already attached for file system {0}"
+                              .format(fs_template['name']))
+                    size = max(fs_template.get('size', 0), fs_template.get('min_size', 0)) # always default to 0 for snaps
+                    if ServiceRole.GALAXY_DATA in ServiceRole.from_string_array(fs_template.get('roles', None)):
+                        size = pss
+                    fs.add_volume(size=size, from_snapshot_id=fs_template['snap_id'])
+                elif 'type' in fs_template:
+                    # TODO: This obviates the need for type archive. Instead, archive should be a volume "data_source"
+                    if 'archive' == fs_template['type'] and 'archive_url' in fs_template:
+                        log.debug("Creating an archive-based ({0}) file system named '{1}' with storage type: {2} and pss: {3}"
+                                  .format(fs_template.get('archive_url'), fs_template['name'], storage_type, pss))
+                        if storage_type == 'volume':
+                            size = max(fs_template.get('size', 10), fs_template.get('min_size', 0))
+                            if ServiceRole.GALAXY_DATA in ServiceRole.from_string_array(
```

Comment:
Sorry, the comment was typed in a hurry, so it's clearly not well worded. What I meant o say is this, - at the moment, there is a special flag called "pss", which applies only to the Galaxy Data filesystem. So in the logic, it's special cased all over the place to accommodate the idea that a user can determine the size of a filesystem. If we can make this notion more general, and make any filesystem's size configurable/determinable by the user.

One way to do this would be to show a UI screen with the filesystems displayed in the launcher page - the filesystems to display can be easily determined by interrogating the flavour data. The user can then select  the desired sizes for each filesystem, or leave the defaults in. We have most of the stuff in place for such a move now, only the UI is missing. (e.g. There is a min_size attribute to each filesystem, which is the minimum allowed size for the filesystem, while the "size" is the user specified size). This doesn't have to be a priority item, just something we can keep in mind.

In the longer term, that will allow the logic as a whole to be moved into a separate ClusterTemplateBuilder or some such class, whose job will be to read in a cluster/filesystem template, and instantiate the relevant components accordingly. Other things that could go into the cluster templates are the desired list of applications to start etc.